x-pack/server/lib/create_router/license_pre_routing_factory/license_pre_routing_factory.js

```diff
@@ -7,21 +7,20 @@
 import { once } from 'lodash';
 import { wrapCustomError } from '../error_wrappers';
 
-export const licensePreRoutingFactory = once((server, pluginId) => {
-  const xpackMainPlugin = server.plugins.xpack_main;
+export const licensePreRoutingFactory = (server, pluginId) => {
+  return () => {
+    return once(() => {
+      // License checking and enable/disable logic
+      const xpackMainPlugin = server.plugins.xpack_main;
+      const licenseCheckResults = xpackMainPlugin.info.feature(pluginId).getLicenseCheckResults();
+      if (!licenseCheckResults.isAvailable) {
+        const error = new Error(licenseCheckResults.message);
+        const statusCode = 403;
+        throw wrapCustomError(error, statusCode);
+      }
 
-  // License checking and enable/disable logic
-  function licensePreRouting() {
-    const licenseCheckResults = xpackMainPlugin.info.feature(pluginId).getLicenseCheckResults();
-    if (!licenseCheckResults.isAvailable) {
-      const error = new Error(licenseCheckResults.message);
-      const statusCode = 403;
-      throw wrapCustomError(error, statusCode);
-    }
-
-    return null;
-  }
-
-  return licensePreRouting;
-});
+      return null;
+    })(server, pluginId);
+  };
+};
```

Comment:
I don't think we need to call with the arguments as they are in the closure.

This is easier to follow for me

```js
export const licensePreRoutingFactory = (server, pluginId) => {
  const licensePreRouting = () => {
    // License checking and enable/disable logic
    const xpackMainPlugin = server.plugins.xpack_main;
    const licenseCheckResults = xpackMainPlugin.info.feature(pluginId).getLicenseCheckResults();
    if (!licenseCheckResults.isAvailable) {
      const error = new Error(licenseCheckResults.message);
      const statusCode = 403;
      throw wrapCustomError(error, statusCode);
    }

    return null;
  }

  return () => once(licensePreRouting)();
};
```

Reply:
I think using `once` here is an issue as the license info can change while Kibana is still running.  There's some logic to reload the xpackInfo when this happens, and using `once` this change would not be observed.

Reply:
Ah I see @bmcconaghy. This was just a refactor but apparently, it helped to spot this! 😊

Reply:
The original did not use `once` AFAICT.

Reply:
Sorry, misunderstood your comment. Indeed the original did not return the licensePreRouting wrapped in `once`.